src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py

```diff
@@ -51,6 +52,38 @@ def str_to_sql(value: str) -> str:
     return f"'{sql_str}'"
 
 
+def str_to_sql_for_year_month_interval(
+    value: str, datatype: YearMonthIntervalType
+) -> str:
+    """
+    Converts "INTERVAL YY-MM [YEAR TO MONTH | YEAR | MONTH]" to quoted format, optimizing based on datatype:
+    - Same start/end field: extracts specific value (YEAR or MONTH only)
+    - Different fields: uses full "YEAR TO MONTH" format
+
+    Examples:
+        "INTERVAL 1-2 YEAR TO MONTH" -> "INTERVAL '1-2' YEAR TO MONTH"
+        "INTERVAL 5-0 YEAR TO MONTH" + YearMonthIntervalType(0,0) -> "INTERVAL '5' YEAR"
+        "INTERVAL 0-3 YEAR TO MONTH" + YearMonthIntervalType(1,1) -> "INTERVAL '3' MONTH"
+    """
+    parts = value.split(" ")
+    if len(parts) < 2:
+        raise ValueError(f"Invalid interval format: {value}")
+    extracted_values = parts[1]
+    start_field = datatype.start_field if datatype.start_field is not None else 0
+    end_field = datatype.end_field if datatype.end_field is not None else 1
+    if datatype.start_field == datatype.end_field:
```

Comment:
what does start_field == end_field imply? Let's add that context here as a comment as well.